### DISCUSSION TITLE
Timestampz date not saved properly

### DISCUSSION THREAD

**Discussion by orisa** [supabase/supabase#14363] [General]
I have a state_date of type timestampz in my database. When I tried saving a date with the value 2023-05-15 12:37:59.611000+0200, it gets saved as 2023-05-15 12:37:59.611+00. Why is the timezone saved as +00. I tried other timezone values same result.

**Comment by vickkhera** [comment]
I think you should check your stored time more carefully, paying particular attention to the hour. Postgres does handle timezones exceptionally well.

```
postgres=*> create table foo (t timestamp with time zone);
CREATE TABLE
Time: 6.825 ms
postgres=*> insert into foo values ('2023-05-15 12:37:59.611000+0200');
INSERT 0 1
Time: 4.649 ms
postgres=*> select * from foo;
             t
----------------------------
 2023-05-15 10:37:59.611+00
(1 row)

Time: 3.420 ms
```

If you want the output of your timestamps in a particular timezone, you can ask Postgres for it. There doesn't appear to be a way to do that with the JS library or the PgREST interface. This is a shame because I really don't want to be doing timezone conversions in my application code, and I cannot set the timezone globally for everyone.

In a direct SQL connection you would do it like this:

```
postgres=*> set local time zone 'America/New_York';
SET
Time: 7.312 ms
postgres=*> select * from foo;
             t
----------------------------
 2023-05-15 06:37:59.611-04
(1 row)
```

or per field in a query like this:

```
postgres=*> select t at time zone 'America/Los_Angeles' from foo;
        timezone
-------------------------
 2023-05-15 03:37:59.611
(1 row)
```

Note in this last case the return type is `timestamp` not `timestamptz`.

**Reply by steve-chavez** [reply]
It's already possible to change timezones with PostgREST. Check here:

https://postgrest.org/en/stable/how-tos/working-with-postgresql-data-types.html#timestamps

Let me know if that works

**Reply by orisa** [reply]
Thanks. I just figured out it converts the saved time to UTC format. I will see how that would affect my app cos  time calculations are involved.

**Reply by vickkhera** [reply]
> https://postgrest.org/en/stable/how-tos/working-with-postgresql-data-types.html#timestamps

"The response has the date in the time zone configured by the server". 

This is just passing the timezone in the timestamp whether it is numeric or symbolic. You cannot make it return the stored timestamp in the timezone you want.

**Comment by steve-chavez** [comment]
Try setting your timezone on the PostgREST role:

```sql
alter role authenticator set time zone 'America/New_York';
```

The above should enforce your timezone across all requests.

Give it a restart just in case. Or try a `NOTIFY pgrst`.

**Reply by vickkhera** [reply]
Not a useful use case for me. I have customer across the country. Their UI should operate in _their_ timezone, and their queries should as well. Things like "show me new items added since midnight (of my timezone)". The list of times in the response should also be in their timezone.

What I *really* need is for there to be a way to pass a header to the REST API for the time zone, and have it run `set local time zone 'America/New_York'` for the duration of that request. This is a feature enhancement, for sure.

**Reply by steve-chavez** [reply]
@vickkhera That seems like a reasonable addition.

It's surprising that there's no standard header for specifying the timezone. Check this discussion around `Accept-Timezone`:

> You are suggesting that this be sent on every HTTP request? Is that correct?
> If you multiply 46 bytes (the line above) up by the number of HTTP
> requests issued across the web every day (billions? trillions?), you can
> imagine how much more data transfer this proposal would require.

https://groups.google.com/g/mozilla.dev.platform/c/hyQQbC3OC-I

Would that be a concern for you?

---

Alternatively, would [custom database roles](https://github.com/orgs/supabase/discussions/11948#discussioncomment-4802137) work for you?

You could map a group of web users into a `la_customers` role and:

```sql
alter role la_customers set time zone 'America/Los_Angeles';
```

**Reply by vickkhera** [reply]
Custom roles would complicate the RLS tremendously, but it is an idea.

As for adding 46 bytes to every Pg REST API call, that's insignificant. I'd be making a DB query to do "set time zone 'America/New_York'" which is probably more with the full protocol round trip. If I could just set it once on `createClient()` and have the JS library pass the header automatically, that'd be even easier for me. But first we gotta get PostGREST to do its part. 🤷🏽

**Reply by steve-chavez** [reply]
@vickkhera Fair points. I've added an issue here: https://github.com/PostgREST/postgrest/issues/2799
